contracts/account/paymaster/PaymasterCore.sol

```diff
@@ -0,0 +1,133 @@
+        return _validatePaymasterUserOp(userOp, userOpHash, maxCost);
+    }
+
+    /**
+     * @dev Internal validation of whether the paymaster is willing to pay for the user operation.
+     * Returns the context to be passed to postOp and the validation data.
+     *
+     * The `requiredPreFund` is the amount the paymaster has to pay (in native tokens). It's calculated
+     * as `requiredGas * userOp.maxFeePerGas`, where `required` gas can be calculated from the user operation
+     * as `verificationGasLimit + callGasLimit + paymasterVerificationGasLimit + paymasterPostOpGasLimit + preVerificationGas`
+     */
+    function _validatePaymasterUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 requiredPreFund
+    ) internal virtual returns (bytes memory context, uint256 validationData);
+
+    /// @inheritdoc IPaymaster
+    function postOp(
+        PostOpMode mode,
+        bytes calldata context,
+        uint256 actualGasCost,
+        uint256 actualUserOpFeePerGas
+    ) public virtual onlyEntryPoint {
+        _postOp(mode, context, actualGasCost, actualUserOpFeePerGas);
+    }
+
+    /**
+     * @dev Handles post user operation execution logic. The caller must be the entry point.
+     *
+     * It receives the `context` returned by `_validatePaymasterUserOp`. Reverts by default
+     * since the function is not called if no context is returned by {validatePaymasterUserOp}.
+     * Overriding without calling `super._postOp(...)` is advised.
+     *
+     * NOTE: The `actualUserOpFeePerGas` is not `tx.gasprice`. A user operation can be bundled with other transactions
+     * making the gas price of the user operation to differ.
+     */
+    function _postOp(
+        PostOpMode /* mode */,
+        bytes calldata /* context */,
+        uint256 /* actualGasCost */,
+        uint256 /* actualUserOpFeePerGas */
+    ) internal virtual {
+        // Must override if `context` is provided
+        revert();
+    }
+
+    /// @dev Calls {IEntryPointStake-depositTo}.
+    function _deposit(address to, uint256 value) internal virtual {
+        entryPoint().depositTo{value: value}(to);
+    }
+
+    /// @dev Variant of {_deposit} that deposits `msg.value` to the paymaster itself (i.e. `address(this)`).
+    function _deposit() internal virtual {
+        _deposit(address(this), msg.value);
+    }
+
+    /// @dev Calls {IEntryPointStake-withdrawTo}.
+    function _withdraw(address payable to, uint256 value) internal virtual {
+        entryPoint().withdrawTo(to, value);
+    }
+
+    /// @dev Calls {IEntryPointStake-addStake}.
+    function _addStake(uint256 value, uint32 unstakeDelaySec) internal virtual {
+        entryPoint().addStake{value: value}(unstakeDelaySec);
+    }
+
+    /// @dev Variant of {_addStake} that uses `msg.value` as the stake amount.
+    function _addStake(uint32 unstakeDelaySec) internal virtual {
+        _addStake(msg.value, unstakeDelaySec);
+    }
```

Comment:
IMO, we shouldn't have that variant. calling the other version with `msg.value` is super easy, and it is clearer/more explicit.